### DISCUSSION TITLE
How can I use 'keepFileExtension' like a babel option ?

### DISCUSSION THREAD

**Discussion by A-312** [babel/babel#11723] [Q&A] [answered]
Hi,

I work on my build function for my `development` env, I try to reduce the building time to <200ms because the production script take 5 sec to build. (My `production` env is ready, I would like to not change the behavior of my prod env, that why I need `keepFileExtension` I would like to not rename my file and change my import).

I have to use [keepFileExtension](https://github.com/babel/babel/pull/6221) option, but I don't see this option in [Babel options](https://babeljs.io/docs/en/options). I need to use this option to keep my original extension (to not change the behavior with my production env). How can I use it ?

I try : 
```js
const buildChatJSX = () =>
  gulp.src(['**/*.jsx'])
    .pipe(babel({
        presets: ['@babel/preset-react'],
        // keepFileExtension: true <--- Options invalid :( How can I use it ?
    }))
    .pipe(gulp.dest('javascript'))
```

But this code is wrong, I can't use `keepFileExtension` :(

> Unknown option: .keepFileExtension. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.

I want to keep `.jsx` extension because, I want to run babel only on `.jsx` files (to earn sometime) and because in my code I did `import render from '../react/main.jsx'`, I can't rename it `js` because this will break my production build script.

Thanks!

----

My package.json : 

```js
  "dependencies": {
    "@babel/core": "7.9.6",
    "@babel/preset-env": "7.9.6",
    "@babel/preset-react": "7.9.4",
```

----

I saw that the option is [here](https://github.com/babel/babel/blob/master/packages/babel-cli/src/babel/options.js#L311) and seems missing [here](https://github.com/babel/babel/blob/master/packages/babel-core/src/config/validation/options.js#L355).

https://github.com/babel/babel/blob/fd3c76941e2e826cacab8844672f1d985f7ea564/packages/babel-cli/src/babel/options.js#L311

https://github.com/babel/babel/blob/fd3c76941e2e826cacab8844672f1d985f7ea564/packages/babel-core/src/config/validation/options.js#L350-L355

**Comment by nicolo-ribaudo** [comment, marked as answer]
It's not possible.

This is because `@babel/core` doesn't have the concept of "file extension": it considers file names as opaque strings, which are matched to your `include`/`exclude` options regardless of their "file path" nature.

File names are handled by Babel integrations, like `@babel/cli`, `babel-loader` or `gulp-babel`: `--keep-file-extension` is an option specific to `@babel/cli`.

`gulp-babel` [hardcodes](https://github.com/babel/gulp-babel/blob/b1f39a8d1d4309d0c3d98194f75acaf5a3ad66a1/index.js#L51) the `.js` extension: I would be happy to review a PR adding a `keepFileExtension` option to that package, but in the meanwhile I suggest using something like [`gulp-rename`](https://www.npmjs.com/package/gulp-rename).
